File: __fixtures__/vuln-samples/vulnerabilities.js
/**
 * Security vulnerability test file — JavaScript
 * 30 intentional vulnerabilities for rule coverage testing.
 * DO NOT USE IN PRODUCTION.
 */

import axios from "axios";

// =================================================================
// 1. SQL Injection — string concat
// =================================================================
function findUser(name) {
  const query = "SELECT * FROM users WHERE name = '" + name + "'";
  return db.query(query);
}

function updateRole(userId, role) {
  const sql = "UPDATE users SET role = '" + role + "' WHERE id = " + userId;
  return db.exec(sql);
}

// =================================================================
// 2. XSS — innerHTML
// =================================================================
function showMessage(msg) {
  document.getElementById("output").innerHTML = msg;
}

function renderSearch() {
  const q = new URLSearchParams(location.search).get("q");
  document.getElementById("results").innerHTML = "<p>" + q + "</p>";
}

// =================================================================
// 3. XSS — document.write
// =================================================================
function legacyRender(data) {
  document.write("<div>" + data + "</div>");
}

// =================================================================
// 4. eval / new Function
// =================================================================
function dynamicCalc(expr) {
  return eval(expr);
}

function buildHandler(code) {
  return new Function("event", code);
}

// =================================================================
// 5. Command injection — exec + execSync
// Intentionally unsafe patterns for scanner testing
// =================================================================
function compress(filename) {
  const { exec } = require("child_process"); // eslint-disable-line
  exec("gzip " + filename, (err, out) => out); // eslint-disable-line
}

function diskUsage(path) {
  const { execSync } = require("child_process"); // eslint-disable-line
  return execSync("du -sh " + path).toString(); // eslint-disable-line
}

// =================================================================
// 6. Path traversal
// =================================================================
function serveFile(req, res) {
  const fs = require("fs");
  const data = fs.readFileSync("/data/" + req.query.name, "utf8");
  res.send(data);
}

function sendDownload(req, res) {
  res.sendFile("/uploads/" + req.params.file);
}

// =================================================================
// 7. SSRF
// =================================================================
async function proxyGet(url) {
  const resp = await axios.get(url);
  return resp.data;
}

async function webhook(target) {
  await fetch(target);
}

async function proxyPost(endpoint, body) {
  await axios.post(endpoint, body);
}

// =================================================================
// 8. Hardcoded secrets
// =================================================================
const AWS_SECRET_KEY = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY";
const AWS_ACCESS_KEY = "AKIAIOSFODNN7EXAMPLE";
const API_KEY = "sk-live-1234567890abcdef";
const JWT_SECRET = "super-secret-jwt-token-never-share";
const PRIVATE_KEY = "-----BEGIN RSA PRIVATE KEY-----\nMIIEow...";

const config = {
  password: "admin123",
};

// =================================================================
// 9. Weak crypto — hash
// =================================================================
const crypto = require("crypto");

function md5Hash(val) {
  return crypto.createHash("md5").update(val).digest("hex");
}

function sha1Hash(val) {
  return crypto.createHash("sha1").update(val).digest("hex");
}

// =================================================================
// 10. Weak crypto — cipher
// =================================================================
function desEncrypt(text, key) {
  const cipher = crypto.createCipheriv("des-ecb", key, null);
  return cipher.update(text, "utf8", "hex") + cipher.final("hex");
}

function rc4Encrypt(text, key) {
  const cipher = crypto.createCipheriv("rc4", key, null);
  return cipher.update(text, "utf8", "hex");
}

// =================================================================
// 11. Insecure random
// =================================================================
function generateOTP() {
  return Math.random().toString().slice(2, 8);
}

function shuffleToken() {
  return Math.random().toString(36).substring(2);
}

// =================================================================
// 12. Open redirect
// =================================================================
function handleRedirect(req, res) {
  const next = req.query.next;
  res.redirect(next);
}

function clientRedirect(url) {
  window.location.href = url;
}

function assignRedirect(dest) {
  window.location = dest;
}

// =================================================================
// 13. Sensitive data in logs
// =================================================================
function authUser(username, password) {
  console.log("Authenticating", password);
  return authenticate(username, password);
}

function chargeCard(creditCard, cvv) {
  console.log(`Charging card ${creditCard} cvv ${cvv}`);
}

// =================================================================
// 14. Sensitive data in localStorage
// =================================================================
function cachePayment(card) {
  localStorage.setItem("lastPayment", JSON.stringify({ cardNumber: card }));
}

function saveToken(token) {
  localStorage.setItem("secret_token", token);
}

// =================================================================
// 15. Prototype pollution
// =================================================================
function merge(target, source) {
  for (const key in source) {
    if (typeof source[key] === "object") {
      target[key] = target[key] || {};
      merge(target[key], source[key]);
    } else {
      target[key] = source[key];
    }
  }
  return target;
}

// =================================================================
// 16. Regex DoS (ReDoS)
// =================================================================
function validateEmail(email) {
  return /^([a-zA-Z0-9_\-\.]+)*@([a-zA-Z0-9_\-\.]+)*\.([a-zA-Z]{2,5})$/.test(email);
}

// =================================================================
// 17. Unvalidated file upload type
// =================================================================
function uploadAvatar(req) {
  const fs = require("fs");
  const path = "/uploads/" + req.file.originalname;
  fs.writeFileSync(path, req.file.buffer);
}

// =================================================================
// 18. Timing-safe comparison missing
// =================================================================
function verifyToken(provided, expected) {
  return provided === expected;  // timing attack
}

// =================================================================
// 19. Insecure cookie
// =================================================================
function setSession(res, token) {
  res.cookie("session", token, { httpOnly: false, secure: false });
}

// =================================================================
// 20. Hardcoded IP / internal URL
// =================================================================
const INTERNAL_API = "http://192.168.1.100:8080/api";
const ADMIN_URL = "http://10.0.0.1:3000/admin";

export {
  findUser, showMessage, dynamicCalc, compress, serveFile,
  proxyGet, md5Hash, desEncrypt, generateOTP, handleRedirect,
  authUser, cachePayment, merge, validateEmail, verifyToken,
};

File: __fixtures__/vuln-samples/vulnerabilities.py
"""
Security vulnerability test file — Python
30 intentional vulnerabilities for rule coverage testing.
DO NOT USE IN PRODUCTION.

NOTE: All vulnerabilities are INTENTIONAL for security scanner validation.
"""

import os
import hashlib
import pickle  # nosec — intentional for testing
import subprocess  # nosec — intentional for testing
import sqlite3
import random
import yaml
import xml.etree.ElementTree as ET
import requests as http_requests
from flask import Flask, request, redirect, render_template_string, jsonify

app = Flask(__name__)

# =================================================================
# 1. SQL Injection — string concat + f-string
# =================================================================
def get_user(username):
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM users WHERE name = '" + username + "'")
    return cursor.fetchall()

def delete_user(user_id):
    conn = sqlite3.connect("app.db")
    conn.execute(f"DELETE FROM users WHERE id = {user_id}")
    conn.commit()

def search(keyword):
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    cursor.execute(f"SELECT * FROM posts WHERE body LIKE '%{keyword}%'")
    return cursor.fetchall()

# =================================================================
# 2. XSS — template injection
# =================================================================
@app.route("/hello")
def hello():
    name = request.args.get("name", "")
    return render_template_string(f"<h1>Hello {name}</h1>")

@app.route("/preview")
def preview():
    html = request.args.get("html", "")
    return render_template_string(html)

# =================================================================
# 3. Command injection — os.system + subprocess shell=True
# All intentional for scanner testing (nosec)
# =================================================================
def ping(host):
    os.system("ping -c 4 " + host)  # nosec

def grep_logs(pattern):
    output = subprocess.check_output(  # nosec
        "grep " + pattern + " /var/log/app.log", shell=True
    )
    return output

def tail_file(filename):
    result = subprocess.Popen(  # nosec
        f"tail -100 {filename}",
        shell=True,
        stdout=subprocess.PIPE,
    )
    return result.stdout.read()

def run_script(name):
    subprocess.run("bash /scripts/" + name, shell=True)  # nosec

def call_tool(cmd):
    subprocess.call(cmd, shell=True)  # nosec

# =================================================================
# 4. Hardcoded secrets
# =================================================================
SECRET_KEY = "flask-secret-key-do-not-share"
DATABASE_URL = "postgresql://admin:p@ssw0rd@prod-db:5432/app"
AWS_ACCESS_KEY_ID = "AKIAIOSFODNN7EXAMPLE"
AWS_SECRET_ACCESS_KEY = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
STRIPE_SECRET_KEY = "stripe_secret_EXAMPLE_DO_NOT_USE"
SENDGRID_API_KEY = "SG.xxxxxxxxxxxxxxxxxxxx"
PRIVATE_KEY = "-----BEGIN RSA PRIVATE KEY-----\nMIIEow..."

app.config["SECRET_KEY"] = "hardcoded-123"

# =================================================================
# 5. Insecure deserialization (all intentional — nosec)
# =================================================================
def load_session(cookie_data):
    import base64
    return pickle.loads(base64.b64decode(cookie_data))  # nosec

def parse_yaml(raw):
    return yaml.load(raw, Loader=yaml.Loader)  # nosec

def parse_unsafe_yaml(raw):
    return yaml.unsafe_load(raw)  # nosec

# =================================================================
# 6. Path traversal
# =================================================================
@app.route("/download")
def download():
    filename = request.args.get("file")
    with open("/uploads/" + filename, "rb") as f:
        return f.read()

@app.route("/avatar/<username>")
def avatar(username):
    return open(f"/data/avatars/{username}.png", "rb").read()

# =================================================================
# 7. SSRF
# =================================================================
@app.route("/fetch")
def fetch_url():
    url = request.args.get("url")
    resp = http_requests.get(url)
    return resp.text

@app.route("/webhook")
def send_webhook():
    target = request.json["callback"]
    http_requests.post(target, json={"ok": True})
    return "sent"

@app.route("/proxy")
def proxy():
    dest = request.args.get("dest")
    http_requests.put(dest, data=request.get_data())
    return "ok"

# =================================================================
# 8. Weak cryptography
# =================================================================
def hash_pw(password):
    return hashlib.md5(password.encode()).hexdigest()

def hash_sha1(data):
    return hashlib.sha1(data.encode()).hexdigest()

def weak_token(user_id):
    return hashlib.md5(f"{user_id}-token".encode()).hexdigest()

# =================================================================
# 9. Insecure random
# =================================================================
def otp():
    return random.randint(100000, 999999)

def session_id():
    chars = "abcdef0123456789"
    return "".join(random.choice(chars) for _ in range(32))

def coin_flip():
    return random.random() > 0.5

# =================================================================
# 10. Open redirect
# =================================================================
@app.route("/login-redirect")
def login_redirect():
    next_url = request.args.get("next", "/")
    return redirect(next_url)

@app.route("/oauth/callback")
def oauth_callback():
    return_to = request.args.get("return_to")
    return redirect(return_to)

# =================================================================
# 11. Sensitive data exposure — print/log
# =================================================================
def authenticate(username, password):
    print(f"Auth: user={username}, pass={password}")
    return True

def process_payment(card, cvv):
    print(f"Payment: card={card}, cvv={cvv}")

# =================================================================
# 12. Debug / environment exposure
# =================================================================
@app.route("/debug")
def debug():
    return jsonify({
        "env": dict(os.environ),
        "config": str(app.config),
    })

# =================================================================
# 13. Mass assignment
# =================================================================
def update_profile(user_id, data):
    conn = sqlite3.connect("app.db")
    for key, val in data.items():
        conn.execute(f"UPDATE users SET {key} = '{val}' WHERE id = {user_id}")
    conn.commit()

# =================================================================
# 14. XXE
# =================================================================
def parse_xml(raw):
    return ET.fromstring(raw)

# =================================================================
# 15. Timing attack
# =================================================================
def check_token(provided, stored):
    return provided == stored  # not timing-safe

# =================================================================
# 16. CORS wildcard
# =================================================================
@app.after_request
def add_cors(response):
    response.headers["Access-Control-Allow-Origin"] = "*"
    return response
